src/nbl/asset/utils/CGLSLCompiler.cpp

```diff
@@ -0,0 +1,155 @@
+{
+    core::smart_refctd_ptr<ICPUBuffer> outSpirv = nullptr;
+    if (code)
+    {
+        core::smart_refctd_ptr<asset::ICPUShader> cpuShader;
+        if (options.includeFinder != nullptr)
+        {
+            cpuShader = resolveIncludeDirectives(code, options.stage, options.sourceIdentifier, options.maxSelfInclusionCount, options.logger);
+            if (cpuShader)
+            {
+                code = reinterpret_cast<const char*>(cpuShader->getContent()->getPointer());
+            }
+        }
+
+        if (strcmp(options.entryPoint, "main") == 0)
+        {
+            shaderc::Compiler comp;
+            shaderc::CompileOptions shadercOptions; //default options
+            shadercOptions.SetTargetSpirv(static_cast<shaderc_spirv_version>(options.targetSpirvVersion));
+            const shaderc_shader_kind stage = options.stage == IShader::ESS_UNKNOWN ? shaderc_glsl_infer_from_source : ESStoShadercEnum(options.stage);
+            const size_t glsl_len = strlen(code);
+            if (options.genDebugInfo)
+                shadercOptions.SetGenerateDebugInfo();
+
+            shaderc::AssemblyCompilationResult asm_res;
+            shaderc::SpvCompilationResult bin_res;
+            if (options.outAssembly)
+            {
+                asm_res = comp.CompileGlslToSpvAssembly(code, glsl_len, stage, options.sourceIdentifier ? options.sourceIdentifier : "", shadercOptions);
+                options.outAssembly->resize(std::distance(asm_res.cbegin(), asm_res.cend()));
+                memcpy(options.outAssembly->data(), asm_res.cbegin(), options.outAssembly->size());
+                bin_res = comp.AssembleToSpv(options.outAssembly->data(), options.outAssembly->size(), shadercOptions);
+            }
+            else
+            {
+                bin_res = comp.CompileGlslToSpv(code, glsl_len, stage, options.sourceIdentifier ? options.sourceIdentifier : "", options.entryPoint, shadercOptions);
+            }
+
+            if (bin_res.GetCompilationStatus() == shaderc_compilation_status_success)
+            {
+                outSpirv = core::make_smart_refctd_ptr<ICPUBuffer>(std::distance(bin_res.cbegin(), bin_res.cend()) * sizeof(uint32_t));
+                memcpy(outSpirv->getPointer(), bin_res.cbegin(), outSpirv->getSize());
+
+                if (options.spirvOptimizer)
+                    outSpirv = options.spirvOptimizer->optimize(outSpirv.get(), options.logger);
+            }
+            else
+            {
+                options.logger.log(bin_res.GetErrorMessage(), system::ILogger::ELL_ERROR);
+            }
+        }
+        else
+        {
+            options.logger.log("shaderc requires entry point to be \"main\" in GLSL", system::ILogger::ELL_ERROR);
+        }
+    }
+    else
+    {
+        options.logger.log("code is nullptr", system::ILogger::ELL_ERROR);
+    }
+    return outSpirv;
```

Comment:
minor stylistic issue, to avoid "Python Levels of Indentation" esp if a function returns a pointer (dumb or smart) its better to do 
```cpp
if (failCondition)
{
   log("the what and why", system::ILogger::ELL_ERROR);
   return nullptr;
}
   
// rest of code
```

Reply:
yea I hoped you wouldn't notice that, cause it's a syntax I use on my own.
I prefer the this way cause all the failure points can be viewed next to each other and there is only a single point of return which can be inspected while debugging.
I'll change the style to your preference though.